### ISSUE TITLE
PickleDataset is not deepcopiable for kedro>=0.17.6

### ISSUE THREAD

**Issue by Galileo-Galilei** [kedro-org/kedro#1134]
## Description

``PickleDataset`` was  modified in ``kedro==0.17.6`` to support any arbitrary backend.

The backend module is now stored as an attribute of the dataset: 

https://github.com/quantumblacklabs/kedro/blob/319a9177c45fac4fe55a0805bd2d83e6d4fa59f6/kedro/extras/datasets/pickle/pickle_dataset.py#L178

## Context

I tried to deepcopy the catalog  (in order to use it in a custom mlflow model : https://github.com/Galileo-Galilei/kedro-mlflow/blob/master/kedro_mlflow/mlflow/kedro_pipeline_model.py#L204), and got the following error : ``TypeError: can't pickle module objects``. 

After investigation, it turns out that the error comes from the ``PickleDataSet`` which is no longer deepcopiable.

## Steps to Reproduce
1. Install kedro 0.17.6 in a new environment  ``conda create -n bug_dc python=3.7 kedro==0.17.6 -y``
2. ``conda activate bug_dc``
3. Run this script:
```python
from kedro.extras.datasets.pickle import PickleDataSet
from copy import deepcopy

pkl_ds = PickleDataSet(
    filepath=r"path/to/my_data.pkl"
)
pkl_ds2=deepcopy(pkl_ds) # raise the error
pkl_ds3=pkl_ds._copy() # same error with the builtin function
```

## Expected Result

The dataset should be deecopied.

## Actual Result

Get the following error
```python 
TypeError: can't pickle module objects
```

## Your Environment

* Kedro version used (`pip show kedro` or `kedro -V`): 0.17.6
* Python version used (`python -V`): 3.7.10
* Operating system and version: Windows 10

## Potential solutions

- For now, I will use the built-in ``DataCatalog.shallow_copy()`` function, but this is not a deepcopy and the underlying datasets have the same memory id.
- It may be useful to make _backend a property so that the import is made on the fly when using the attributes, rather than storing the entire module,e.g. :

````python

class PickleDataSet(AbstractVersionedDataSet):

    # self._backend is the module, self.backend is the string
    @property
    def _backend(self):
        # move the import here and remove it from __init__
        try:
            imported_backend = importlib.import_module(self.backend) # self.backend is the string
        except ImportError as exc:
            raise ImportError(
                f"Selected backend '{backend}' could not be imported. "
                "Make sure it is installed and importable."
            ) from exc

        if not (
            hasattr(imported_backend, "load") and hasattr(imported_backend, "dump")
        ):
            raise ValueError(
                f"Selected backend '{backend}' should satisfy the pickle interface. "
                "Missing one of `load` and `dump` on the backend."
            )
    return imported_backend

**Comment by datajoely** [comment]
Hi @Galileo-Galilei thanks for raising this - I can't speak for the dev team but your potential solution looks sensible so if you have a chance to raise a PR I'm pretty sure it will get merged quickly. We always err on the side of being lazy when we can - @mzjp2 do you see any issues with this approach?

**Comment by Galileo-Galilei** [comment]
Hi @datajoely, I'll give it a try hopefully this week, but I had a lot of troubles running the test suite on windows in the past and I have little time for now, so i am not sure of the delay. 

Do you think we need to make a call to the property (call ``self._backend``) in ``__init__`` to ensure an error is raised if a module is missing at initialisation instead of at loading time? Above change would raise an ``ImportError`` at loading time, which can occur much further in the pipeline than it does for now (the error is raised as soon as the catalog is created with current implementation).

**Comment by johnkmaxi** [comment]
Is this related to issues with running a pipline using the ParallelRunner? I get this error when running a pipeline using the parallel flag. 
kedro version 0.17.6
python version 3.7.11
windows server 2016

**Comment by datajoely** [comment]
@johnkmaxi that's very good point - if your pipeline is using the `PickleDataSet` and you try and use `ParallelRunner` it may cause issues. We're prioritising a fix for this as we speak.

**Comment by johnkmaxi** [comment]
Great to hear. Thank you @datajoely 

In the interim, I was able to utilize multiprocessing to parallelize my code execution by using multiprocces inside of my nodes themselves. `mutliprocess_wrapper` and `model_evaluator` live in pipelines.data_science.common and are imported into the lower level node associated with that model. `type_of_model` is a node function that accepts inputs and packages them into the `multiprocess_wrapper` and a node level `_type_of_model` that does the actual model training. 

Looking forward to a fix, as this is a serviceable, but not great alternative when using pickle datasets.

in a module holding common functions for use in any node:
```python
def multiprocess_wrapper(*inputs, model_fn=None):
    """Train Timeseries model

    Parameters
    ----------
    *inputs: tuple
        series: pd.Series
        model_func: callable
        params: dict, general parameters
        model_fn: callable

    Returns
    -------
    list: [results forecasts]*(len(inputs) - 2)
        results, a pd.DataFrame, of performance per h
        forecasts, a darts.TimeSeries, of predictions used to calculate the results
    """
    datasets = inputs[:-2]
    params = inputs[-2]
    model_params = inputs[-1]
    data = list(
        zip(
            [model_fn for x in range(len(datasets))],
            datasets,
            [params for x in range(len(datasets))],
            [model_params for x in range(len(datasets))],
        )
    )
    pool = multiprocessing.pool.Pool(processes=params["n_processes"])
    # model_evaluator is function to evaluate arbitrary model
    result = pool.starmap(model_evaluator, data)
    # result is list of lists that needs to be flattened
    # [[data1 performance, data1 forecast], [data2 performance, data2 forecast]]
    # -> [data1 performance, data1 forecast, data2 performance, data2 forecast]
    return [item for sublist in result for item in sublist]
```
In a node module:
```python
def type_of_model(*inputs):
    """Train Timeseries model

    Parameters
    ----------
    *inputs: tuple
        series: pd.Series
        model_func: callable
        params: dict, general parameters
        model_params: dict, medel specific parameters

    Returns
    -------
    list: [results forecasts]*(len(inputs) - 2)
        results, a pd.DataFrame, of performance per h
        forecasts, a darts.TimeSeries, of predictions used to calculate the results
    """
    return multiprocess_wrapper(*inputs, model_fn=_type_of_model)
```

**Comment by datajoely** [comment]
@johnkmaxi that's a very clever solution - the other option I thought of was to copy the older `PickleDataSet` [implementation from 0.17.5](https://kedro.readthedocs.io/en/0.17.5/_modules/kedro/extras/datasets/pickle/pickle_dataset.html#PickleDataSet) into your project and reference it as a custom dataset.

**Comment by johnkmaxi** [comment]
@datajoely  That is way better way to do it! Still learning the ins and outs of kedro. I think it is a great tool.
Now that I can use ParallelRunner, do you know how to change the "NUMEXPR_MAX_THREADS" variable?
I'm on windows server 2016. I added it as an environment variable via the control panel, but it didn't seem to affect anything when running the pipline.

**Comment by datajoely** [comment]
Hi @johnkmaxi that variable is for numpy so it's not really Kedro, the docs suggest you can do it both by the env variable and dynamically using the python command:

```python
import numexpr
numexpr.set_num_threads(int)
```
https://numexpr.readthedocs.io/projects/NumExpr3/en/latest/user_guide.html#threadpool-configuration

**Comment by AntonyMilneQB** [comment]
Thanks very much for this excellent bug report and suggested solution @Galileo-Galilei. I've fixed it in #1213, which will be released with 0.17.7 - do have a look at the PR if you get a change 🙂  This will also fix the problem with the parallel runner @johnkmaxi.
